aptos-move/framework/aptos-framework/sources/SharedAccount.move

```diff
@@ -0,0 +1,153 @@
+    use AptosFramework::SimpleMap::{Self, SimpleMap};
+    use AptosFramework::ResourceAccount;
+
+    struct Shares has store {
+        numerator: u64,
+        denominator: u64,
+    }
+
+    // Resource representing a shared account
+    struct SharedAccount has key {
+        share_record: SimpleMap<address, Shares>,
+    }
+
+    const EINVALID_INIT_INPUT: u64 = 0;
+    const EACCOUNT_DNE: u64 = 1;
+    const EADDRESS_ALREADY_EXISTS: u64 = 2;
+    const EINVALID_NUMERATOR_DENOMINATOR_COMBINATIONS: u64 = 3;
+    const EINVALID_SIGNER: u64 = 4;
+    const EINSUFFICIENT_BALANCE: u64 = 5;
+
+    public(script) fun initialize_shared_account(source: &signer, seed: vector<u8>): (signer) {
+        assert!(Account::exists_at(Signer::address_of(source)), Errors::invalid_argument(EACCOUNT_DNE));
+        let resource_signer = ResourceAccount::create_resource_account(source, seed, Vector::empty());
+        if (!exists<SharedAccount>(Signer::address_of(&resource_signer))) {
+            move_to(
+                &resource_signer,
+                SharedAccount {
+                    share_record: SimpleMap::create(),
+                }
+            )
+        };
+        resource_signer
+    }
+
+    public(script) fun initialize(source: &signer, seed: vector<u8>, addresses: vector<address>, numerators: vector<u64>, denominator: u64): (signer) acquires SharedAccount {
+        assert!(Vector::length(&addresses) == Vector::length(&numerators), Errors::invalid_argument(EINVALID_INIT_INPUT));
+
+        let resource_signer = initialize_shared_account(source, seed);
+
+        let shared_account = borrow_global_mut<SharedAccount>(Signer::address_of(&resource_signer));
+
+        let sum: u64 = 0;
+        let i = 0;
+
+        while (i < Vector::length(&numerators)) {
+            let num = *Vector::borrow(&numerators, i);
+            let addr = *Vector::borrow(&addresses, i);
+            assert!(Account::exists_at(addr), Errors::invalid_argument(EACCOUNT_DNE));
+            sum = sum + num;
+            i = i + 1;
+
+            SimpleMap::add(&mut shared_account.share_record, addr, Shares { numerator: num, denominator: denominator });
+        };
+        assert!(sum == denominator, Errors::invalid_argument(EINVALID_NUMERATOR_DENOMINATOR_COMBINATIONS));
+        resource_signer
+    }
+
+    public(script) fun disperse<CoinType>(resource_signer: &signer, total_amount: u64) acquires SharedAccount {
+        assert!(exists<SharedAccount>(Signer::address_of(resource_signer)), Errors::invalid_argument(EINVALID_SIGNER));
+        assert!(Coin::balance<CoinType>(Signer::address_of(resource_signer)) >= total_amount, Errors::limit_exceeded(EINSUFFICIENT_BALANCE));
+                
+        let i = 0;
+        let shared_account = borrow_global<SharedAccount>(Signer::address_of(resource_signer));
+
+        while (i < SimpleMap::length(&shared_account.share_record)) {
+            let (key, value) = SimpleMap::get_entry(&shared_account.share_record, i);
```

Comment:
Nevermind about getting the whole vector for the map in my other comment, this is really a pretty good experience here, and is very legible.